File: py/1766.py
def main():
    t = int(input())

    for count in range(t):
        n, m = [int(i) for i in input().split()]
        names = []
        
        for _ in range(n):
            s, p, i, a = input().split()
            p, i, a = int(p), int(i), float(a)
            names.append((s, p, i, a))

        names.sort(key=lambda a: (a[2], a[3], a[0]))
        names.sort(key=lambda a: a[1], reverse=True)
        print('CENARIO', '{' + str(count + 1) + '}')

        for j in range(m):
            print(f'{j + 1} - {names[j][0]}')


main()

File: cpp/1766.cpp
#include <iostream>
#include <string>
#include <vector>
#include <tuple>
#include <algorithm>

using namespace std;

bool sort_criteria(const tuple<string, int, int, float> &a,
                   const tuple<string, int, int, float> &b)
{
    if (get<1>(a) != get<1>(b)) return (get<1>(a) > get<1>(b));
    if (get<2>(a) != get<2>(b)) return (get<2>(a) < get<2>(b));
    if (get<3>(a) != get<3>(b)) return (get<3>(a) < get<3>(b));
    return (get<0>(a) < get<0>(b));
}

int main()
{
    int t, c;
    cin >> t;

    for (c = 1; c <= t; c++) {
        int n, m;
        cin >> n >> m;
        vector<tuple<string, int, int, float>> names;

        for (n; n > 0; n--) {
            string s;
            int p, i;
            float a;
            cin >> s >> p >> i >> a;
            names.push_back(make_tuple(s, p, i, a));
        }

        sort(names.begin(), names.end(), sort_criteria);
        cout << "CENARIO {" << c << "}\n";
        int j = 1;

        for (auto name: names) {
            cout << j << " - " << get<0>(name) << '\n';
            j++;
            if (j > m) break;
        }
    }

    return 0;
}